plugins/ccgx/fu-ccgx-pure-hid-device.c

```diff
@@ -0,0 +1,538 @@
+
+#define ROW_SIZE 128
+#define FW1_START 0x0030
+#define FW2_START 0x0200
+#define FW1_METADATA 0x03FF
+#define FW2_METADATA 0x03FE
+#define WRITE_HEADER_SIZE 4
+
+/**
+ * FU_CCGX_PURE_HID_DEVICE_IS_IN_RESTART:
+ *
+ * Device is in restart and should not be closed manually.
+ *
+ * Since: 1.7.0
+ */
+#define FU_CCGX_PURE_HID_DEVICE_IS_IN_RESTART (1 << 0)
+
+struct _FuCcgxPureHidDevice {
+	FuHidDevice parent_instance;
+	FuCcgxFwMode operating_mode;
+	guint32 versions[FU_CCGX_FW_MODE_LAST];
+	//guint8 bootloader_info;
+	//guint8 bootmode_reason;
+	guint32 silicon_id;
+	//guint32 image_1_row;
+	//guint32 image_2_row;
+	//device_uid: // [u8; 6],
+	guint32 flash_row_size;
+	guint32 flash_size;
+};
+G_DEFINE_TYPE(FuCcgxPureHidDevice, fu_ccgx_pure_hid_device, FU_TYPE_HID_DEVICE)
+
+
+typedef struct __attribute__((packed)) {
+	guint8 report_id;
+	guint8 reserved_1;
+	guint16 signature;
+	guint8 operating_mode;
+	guint8 bootloader_info;
+	guint8 bootmode_reason;
+	guint8 reserved_2;
+	guint32 silicon_id;
+	guint32 bl_version;
+	guint8 bl_version_reserved[4];
+	guint32 image_1_ver;
+	guint8 image_1_ver_reserved[4];
+	guint32 image_2_ver;
+	guint8 image_2_ver_reserved[4];
+	guint32 image_1_row;
+	guint32 image_2_row;
+	guint8 device_uid[6];
+	guint8 reserved_3[10];
+} HidFwInfo;
+
+#define FU_CCGX_PURE_HID_DEVICE_TIMEOUT     5000 /* ms */
+#define FU_CCGX_PURE_HID_DEVICE_RETRY_DELAY 30   /* ms */
+#define FU_CCGX_PURE_HID_DEVICE_RETRY_CNT   5
+
+static gboolean
+fu_ccgx_pure_hid_command(FuDevice *device,
+			guint8 param1,
+			guint8 param2,
+			GError **error)
+{
+	guint8 buf[8] = {CCGX_HID_COMMAND_E1, param1, param2, 0x00, 0xCC, 0xCC, 0xCC, 0xCC};
+
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(device),
+				      buf[0],
+				      buf,
+				      sizeof(buf),
+				      FU_CCGX_PURE_HID_DEVICE_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_NONE,
+				      error)) {
+		return FALSE;
+	}
+	return TRUE;
+}
+
+
+static gboolean
+fu_ccgx_pure_hid_flashing_mode(FuDevice *device, gpointer user_data, GError **error)
+{
+	g_debug("before flashing mode");
+	if (!fu_ccgx_pure_hid_command(device, CCGX_HID_CMD_FLASH, CY_PD_ENTER_FLASHING_MODE_CMD_SIG, error)) {
+		g_prefix_error(error, "flashing enable command error: ");
+		return FALSE;
+	}
+	g_debug("After flashing mode");
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_pure_hid_magic_unlock(FuDevice *device, gpointer user_data, GError **error)
+{
+	guint8 magic_buf[8] = {CCGX_HID_CUSTOM, 0x42, 0x43, 0x59, 0x00, 0x00, 0x00, 0x0B};
+	gboolean err;
+	g_debug("magic unlock");
+
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(device),
+				      magic_buf[0],
+				      magic_buf,
+				      sizeof(magic_buf),
+				      FU_CCGX_PURE_HID_DEVICE_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_IS_FEATURE,
+				      error)) {
+		g_prefix_error(error, "magic enable command error: ");
+		return FALSE;
+	}
+
+
+	g_debug("bridge mode");
+	err = fu_ccgx_pure_hid_command(device, CCGX_HID_CMD_MODE, CY_PD_BRIDGE_MODE_CMD_SIG, NULL);
```

Comment:
just do a g_debug() with a g_autoptr(GError) error_local -- it might be useful for debugging.